### ISSUE TITLE
openconnect 9.01 disconnect issue

### ISSUE THREAD

**Issue by Wongboo** [Homebrew/homebrew-core#100849]
### `brew gist-logs <formula>` link OR `brew config` AND `brew doctor` output

```shell
HOMEBREW_VERSION: 3.4.10-32-g6146564
ORIGIN: https://mirrors.ustc.edu.cn/brew.git
HEAD: 6146564fffdba36dfc126ede49f4a82742aaa43c
Last commit: 12 hours ago
Core tap ORIGIN: https://mirrors.ustc.edu.cn/homebrew-core.git
Core tap HEAD: 48e08538b1015026ff9346fd83a6679cdbcb8e4b
Core tap last commit: 4 hours ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_BOTTLE_DOMAIN: https://mirrors.ustc.edu.cn/homebrew-bottles
HOMEBREW_CASK_OPTS: []
HOMEBREW_CORE_GIT_REMOTE: https://github.com/Homebrew/homebrew-core
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.FAm9TVwBDg/org.xquartz:0
HOMEBREW_EDITOR: code -w
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.8 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit kabylake
Clang: 13.1.6 build 1316
Git: 2.36.0 => /usr/local/bin/git
Curl: 7.79.1 => /usr/bin/curl
macOS: 12.3.1-x86_64
CLT: 13.3.1.0.1.1648687083
Xcode: N/A
```

### Verification

- [X] I ran `brew update` and am still able to reproduce my issue.
- [X] I have resolved all warnings from `brew doctor` and that did not fix my problem.

### What were you trying to do (and why)?

use openconnect after homebrew upgrade it 

### What happened (include all command output)?

I can run `openconnect` fine, vpn works fine, but I cannot disconnect connection, it shows
```
Failed to spawn script /usr/local/etc/vpnc-script for disconnect: Interrupted system call
```

### What did you expect to happen?

disconnect fine

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
brew install openconnect
sudo openconnect --protocol=nc <your server>
press ctrl+c no use, press ctrl+c to force quit, but network has already broken
```

**Comment by SMillerDev** [comment]
Did you try to report this to the developers of openconnect? This doesn't seem specific to Homebrew

**Comment by Wongboo** [comment]
ok, I will

**Comment by cho-m** [comment]
Also may want to make sure `/usr/local/etc/vpnc-script` is okay (exists, permissions okay, etc) and perhaps check if any noticeable differences with `/usr/local/etc/vpnc-script.default` (newer copy) since we do not overwrite existing file. Can also try overriding with `--script` input to try if same issue with newer copy.

**Comment by Wongboo** [comment]
Here is difference and permissions, still couldn't get any clues from output 
```
diff /usr/local/etc/vpnc-script /usr/local/etc/vpnc-script.default                                                                              ─╯
158c158,159
< grep '^hosts' /etc/nsswitch.conf 2>/dev/null|grep resolve >/dev/null 2>&1 && command systemd-resolve --status >/dev/null 2>&1
---
> RESOLVEDENABLED=0
> grep '^hosts' /etc/nsswitch.conf 2>/dev/null|grep resolve >/dev/null 2>&1
160,162c161,164
< 	RESOLVEDENABLED=1
< else
< 	RESOLVEDENABLED=0
---
> 	command resolvectl status >/dev/null 2>&1 || command systemd-resolve --status >/dev/null 2>&1
> 	if [ $? = 0 ];then
> 		RESOLVEDENABLED=1
> 	fi
```
and
```
ls -l  /usr/local/etc/vpnc-script /usr/local/etc/vpnc-script.default                                                                            ─╯
-rwxr-xr-x  1 happy  admin  38346  3 24 18:47 /usr/local/etc/vpnc-script
-rwxr-xr-x  1 happy  admin  38408  5  5 13:58 /usr/local/etc/vpnc-script.default
```

**Comment by dwmw2** [comment]
> Did you try to report this to the developers of openconnect? This doesn't seem specific to Homebrew

Thanks, I think this is https://gitlab.com/openconnect/openconnect/-/issues/423

I think you're right that it isn't specific to Homebrew, but it probably is specific to the OSX build. Sadly we do not have automated testing and CI for the OSX builds.

I doubt the vpnc-script changes are relevant; it'll be something we changed in openconnect. Can you help me to understand the failure mode a bit more please? What does hitting Ctrl-C twice do? Normally in a POSIX environment, hitting Ctrl-C will deliver a SIGINT to the process group. And doing it again would... do it again. In OSX does the second one turn into SIGQUIT or something?

If you could please reproduce this with `-vv` on the command line, that would be much appreciated.

SIGQUIT would abort openconnect *without* attempting to run the script for disconnect, and would leave stray network/DNS configuration around. I think the utun interface itself would disappear, and would take its routing config with it, but perhaps not DNS? Please could you show your routing and DNS config both before, *during* and after the failing disconnect, so we can see what's going on?

We can continue this in the openconnect issue linked above if you prefer, but I wanted to ask some of those questions here where there's a little more OSX clue to be had :)

Thanks!

**Comment by dwmw2** [comment]
Oh, and are you logging in remotely with SSH/remote desktop to do this? If so, and if the VPN 'eats' the route to the client you're logging in *from*, that would explain why Ctrl-C doesn't work.... doesn't cover why a second one would, but I can file that under "things I didn't understand correctly"...

**Comment by dwmw2** [comment]
@alechemy Apologies if you felt unwelcomed on the GitLab issue; your contribution very much *is* welcome. I think you reported that the VPN wasn't working for you even while you were connected? That may or may not be a different issue, so seeing *your* routing+DNS from before/during/after would also be extremely useful. Also, which protocol was that? There *have* been some changes for some VPN protocols around how we handle default routes, and that's my first hypothesis for what might have gone wrong. Thanks!

**Comment by dwmw2** [comment]
(And if anyone has the capacity to do a `git bisect` on openconnect and report precisely where this happened, we can skip a lot of hypothesising and get much more quickly to a solution)

**Comment by Wongboo** [comment]
I'm not logging with ssh, but I'm using ssh to connect another linux remote after connecting to `openconnect` vpn, I think it doesn't matter.
I'm confused that I cannot reproduce the issue today, today I can connect to vpn, but when I press `ctrl+c` for the first time, it outputs to `stdout`:
```
^CGET https://vpn.pku.edu.cn/dana-na/auth/logout.cgi
SSL negotiation with vpn.pku.edu.cn
Connected to HTTPS on vpn.pku.edu.cn with ciphersuite (TLS1.2)-(RSA)-(AES-128-GCM)
Got HTTP response: HTTP/1.1 302 Moved
Set-Cookie: DSID=<elided>; path=/; expires=Thu, 01 Jan 1970 22:00:00 GMT; secure
Set-Cookie: DSLastAccess=x; path=/; expires=Thu, 01 Jan 1970 22:00:00 GMT; secure
Set-Cookie: DSFirstAccess=x; path=/; expires=Thu, 01 Jan 1970 22:00:00 GMT; secure
Set-Cookie: DSRemoteSSOReferer=x; path=/; expires=Thu, 01 Jan 1970 22:00:00 GMT; secure
Set-Cookie: DSRemoteSSOType=x; path=/; expires=Thu, 01 Jan 1970 22:00:00 GMT; secure
Set-Cookie: DSOSMLOGIN=x; path=/; expires=Thu, 01 Jan 1970 22:00:00 GMT; secure
Set-Cookie: DSHCSTARTED=x; path=/dana-na/; expires=Thu, 01 Jan 1970 22:00:00 GMT; secure
Set-Cookie: DSCheckBrowser=x; path=/; expires=Thu, 01 Jan 1970 22:00:00 GMT; secure
Set-Cookie: DSJSAMInitialized=x; path=/; expires=Thu, 01 Jan 1970 22:00:00 GMT; secure
Set-Cookie: DSDID=x; path=/; expires=Thu, 01 Jan 1970 22:00:00 GMT; secure
Set-Cookie: DSLaunchURL=x; path=/; expires=Thu, 01 Jan 1970 22:00:00 GMT; secure
Date: Sat, 07 May 2022 08:26:38 GMT
location: /dana-na/auth/url_default/welcome.cgi?p=logout&u=useruid23779762bf6b1d2ad693ec852ab50dcb750b57ee&signinUrl=IzEjPi12YrazDgABAAAA35jcRzTet5A5bLZT05Yk-bh9bGqVyfnbVlsgqSU6Zq7D%2F-TXpryuAkuDQelNtDQG
ds-connection: close
Content-Type: text/html; charset=utf-8
Pragma: no-cache
Cache-Control: no-store
Expires: -1
Content-Length: 0
X-XSS-Protection: 1
Strict-Transport-Security: max-age=31536000
HTTP body length:  (0)
Logout successful.
route: bad address: 
route: writing to routing socket: File exists
add net default: gateway 192.168.1.2: File exists
delete net 162.105.129.27: gateway 222.29.18.83
delete net 162.105.129.26: gateway 222.29.18.83
route: bad address: 
User cancelled (SIGINT/SIGTERM); exiting.
```
after these output, `openconnect` exits successfully without another `ctrl+c`, but my network seems to be completely broken, might be a route issue?

**Comment by dwmw2** [comment]
Thanks. Yes, this does look like a routing issue. Please could you also show the part of the `-v` output where it's established the connection, and it should print the routes that it is sending to the VPN. Are you able to show the output of `route -n` before/during/after your connection? 

Please make sure you test just with a *numeric* ping or other connection. Let's not get confused by DNS... yet.

**Comment by dwmw2** [comment]
@Wongboo are you able to revert this commit and test, please? https://gitlab.com/openconnect/openconnect/-/commit/faa406bea8702cf8af296ecaf5075f9f730a1399

**Comment by Wongboo** [comment]
@dwmw2 , `openconnect` output
```
GET https://vpn.pku.edu.cn/
Attempting to connect to server 222.29.159.224:443
Connected to 222.29.159.224:443
SSL negotiation with vpn.pku.edu.cn
Connected to HTTPS on vpn.pku.edu.cn with ciphersuite (TLS1.2)-(RSA)-(AES-128-CBC)-(SHA256)
Got HTTP response: HTTP/1.1 302 Found
Location: https://vpn.pku.edu.cn/dana-na/auth/url_default/welcome.cgi
Content-Type: text/html; charset=utf-8
Set-Cookie: DSSIGNIN=url_default; path=/dana-na/; expires=Thu, 31-Dec-2037 00:00:00 GMT; secure
Set-Cookie: DSIVS=; path=/; expires=Thu, 01 Jan 1970 22:00:00 GMT; secure
Set-Cookie: DSSignInURL=/; path=/; secure
Connection: close
Content-Length: 0
HTTP body length:  (0)
GET https://vpn.pku.edu.cn/dana-na/auth/url_default/welcome.cgi
SSL negotiation with vpn.pku.edu.cn
Connected to HTTPS on vpn.pku.edu.cn with ciphersuite (TLS1.2)-(RSA)-(AES-128-CBC)-(SHA256)
Got HTTP response: HTTP/1.1 200 OK
Content-Type: text/html; charset=utf-8
Date: Sat, 07 May 2022 09:12:12 GMT
Pragma: no-cache
Cache-Control: no-store
Expires: -1
Transfer-Encoding: chunked
X-Frame-Options: SAMEORIGIN
HTTP body chunked (-2)
frmLogin
POST https://vpn.pku.edu.cn/dana-na/auth/url_default/login.cgi
Got HTTP response: HTTP/1.1 302 Moved
Set-Cookie: DSASSERTREF=x; path=/; expires=Thu, 01 Jan 1970 22:00:00 GMT; secure
Set-Cookie: DSID=<elided>; path=/; secure
Set-Cookie: DSFirstAccess=1651914732; path=/; secure
Date: Sat, 07 May 2022 09:12:12 GMT
location: https://vpn.pku.edu.cn/dana/home/starter0.cgi?check=yes
Content-Type: text/html; charset=utf-8
Pragma: no-cache
Cache-Control: no-store
Expires: -1
Content-Length: 0
HTTP body length:  (0)
GET https://vpn.pku.edu.cn/dana/home/starter0.cgi?check=yes
Got HTTP response: HTTP/1.1 200 OK
Content-type: text/html; charset=utf-8
Set-Cookie: DSLastAccess=1651914732; path=/; Secure
Pragma: no-cache
Cache-Control: no-store
Expires: -1
Transfer-Encoding: chunked
X-Frame-Options: SAMEORIGIN
HTTP body chunked (-2)
Got HTTP response: HTTP/1.1 200 OK
Content-type: application/octet-stream
Pragma: no-cache
NCP-Version: 3
Set-Cookie: DSLastAccess=1651914732; path=/; Secure
Connection: close
X-Frame-Options: SAMEORIGIN
> 0000:  12 00 00 04 00 00 00 05  00 62 6f 67 6f 6e bb 01  |.........bogon?.|
> 0010:  00 00 00 00                                       |....|
Read 3 bytes of SSL record
< 0000:  01 00 00                                          |...|
Read 358 bytes of SSL record
Got KMP message 301 of length 336
Got KMP message 301 of size 336
Unknown TLV group 3 attr 1 len 1: 01
Unknown TLV group 3 attr 2 len 1: 00
Received split include route 0.0.0.0/0.0.0.0
Received MTU 1400 from server
Received DNS server 162.105.129.27
Received DNS server 162.105.129.26
Received DNS search domain pku.edu.cn
Unknown TLV group 2 attr 3 len 4: 01 00 00 00
Received WINS server 255.255.255.255
ESP compression: 0
ESP encryption: 0x02 (AES-128)
ESP HMAC: 0x02 (SHA1)
ESP key lifetime: 1200 seconds
ESP key lifetime: 0 bytes
ESP replay protection: 1
Unknown TLV group 8 attr 11 len 4: 00 00 00 00
ESP port: 4500
ESP to SSL fallback: 15 seconds
Unknown TLV group 8 attr 8 len 4: 00 00 00 3c
Received internal IP address 222.29.103.44
Received netmask 255.255.255.255
Received internal gateway address 10.110.110.110
ESP SPI (outbound): 98000d0d
64 bytes of ESP secrets
oNCP negotiation request outgoing:
> 0000:  8e 00 00 00 00 00 00 00  01 2f 01 00 00 00 01 00  |........./......|
> 0010:  00 00 00 00 00 10 00 06  00 00 00 0a 00 02 00 00  |................|
> 0020:  00 04 00 00 05 78 00 00  00 00 00 00 01 2e 01 00  |.....x..........|
> 0030:  00 00 01 00 00 00 00 00  00 56 00 07 00 00 00 50  |.........V.....P|
> 0040:  00 01 00 00 00 04 56 d2  73 0e 00 02 00 00 00 40  |......V?s......@|
> 0050:  35 2b be 8c aa 02 a8 44  ff 1a 41 4d 4f 6a 38 ee  |5+?.?.?D?.AMOj8?|
> 0060:  e0 69 fa cc 1d 32 07 1e  86 c5 0e ad 20 1d 50 f9  |?i??.2...?.? .P?|
> 0070:  cf 3d 52 de 00 00 00 00  00 00 00 00 00 00 00 00  |?=R?............|
> 0080:  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
Parameters for incoming ESP: SPI 0x56d2730e
ESP encryption type AES-128-CBC (RFC3602) key 0x352bbe8caa02a844ff1a414d4f6a38ee
ESP authentication type HMAC-SHA-1-96 (RFC2404) key 0xe069facc1d32071e86c50ead201d50f9cf3d52de
Parameters for outgoing ESP: SPI 0x98000d0d
ESP encryption type AES-128-CBC (RFC3602) key 0xf5575b3a97a5f057725b9f3099da47c9
ESP authentication type HMAC-SHA-1-96 (RFC2404) key 0x6d0e4c6d1cf99909ff898603ccd5b364d710820b
Send ESP probes
UDP SO_SNDBUF: 14000
Configured as 222.29.103.44, with SSL connected and ESP in progress
add host 222.29.159.224: gateway 192.168.1.2
route: writing to routing socket: Can't assign requested address
add net 222.29.103.44: gateway 222.29.103.44: Can't assign requested address
delete net default: gateway 192.168.1.2
add net default: gateway 222.29.103.44
add net 162.105.129.27: gateway 222.29.103.44
add net 162.105.129.26: gateway 222.29.103.44
Accepting later-than-expected ESP packet with seq 1 (expected 0)
Received ESP Legacy IP packet of 16 bytes
ESP session established with server
Accepting expected ESP packet with seq 2
Received ESP Legacy IP packet of 16 bytes
Packet outgoing:
> 0000:  21 00 00 00 00 00 00 00  01 2f 01 00 00 00 01 00  |!......../......|
> 0010:  00 00 00 00 00 0d 00 06  00 00 00 07 00 01 00 00  |................|
> 0020:  00 01 01                                          |...|
Sent ESP enable control packet
No work to do; sleeping for 15000 ms...
Sent ESP IPv4 packet of 116 bytes
No work to do; sleeping for 15000 ms...
Accepting expected ESP packet with seq 3
Received ESP Legacy IP packet of 144 bytes
No work to do; sleeping for 15000 ms...
Accepting expected ESP packet with seq 4
Received ESP Legacy IP packet of 704 bytes
No work to do; sleeping for 15000 ms...
Sent ESP IPv4 packet of 84 bytes
No work to do; sleeping for 15000 ms...
Sent ESP IPv4 packet of 116 bytes
Sent ESP IPv4 packet of 116 bytes
No work to do; sleeping for 15000 ms...
Accepting expected ESP packet with seq 5
Received ESP Legacy IP packet of 64 bytes
No work to do; sleeping for 15000 ms...
Sent ESP IPv4 packet of 100 bytes
Sent ESP IPv4 packet of 612 bytes
```
But I cannot get output of `route -n`, probably becasue macOS's route is different from linux
```
route -n                                                                                                                                       
usage: route [-dnqtv] command [[modifiers] args]
```

**Comment by dwmw2** [comment]
> Received split include route 0.0.0.0/0.0.0.0

That really looks like we have an explicit "split include" for the default route, 0.0.0.0/0. And vpnc-script on OSX isn't handling that correctly; leaving your network in a bad state.

For background: `vpnc-script` is a separate tool, which I inherited from the older `vpnc` IPSec client. It is supposed to handle the network configuration and routing on various platforms, leaving openconnect to concentrate on the job of just pushing packets around. This is the model that vpnc used, and it just made sense at the time to use the *same* script and give it the same inputs to ask it to configure networks for openconnect too.

We pass the configuration information that we receive from the VPN server, through environment variables. These are documented (mostly) at the top of vpnc-script itself. The split-include/split-exclude handling is a bit weird but basically works like this:

 * *if* any 'includes' are provided, then those are the only routes configured for the VPN interface.
 * else, *if* any excludes are provided, then the VPN interface has the route for anything *except* those excludes, which still go via the local physical network
 * else, if neither includes or excludes are provided, the VPN just gets the default route.

I think that OpenConnect 8.20 used to use the last bullet point; providing no explicit includes/excludes and so all traffic went to the VPN via the default route (apart from your local physical subnet which will have a higher priority route of its own).

In OpenConnect 9.01 you use the first bullet point, with an *explicit* default route. This really ought to have the *same* effect, but for some reason it isn't.

Are you able to add `set -x` near the top of vpnc-script and show output of both connection and disconnection, from both OpenConnect 8.20 and 9.01 please?

**Comment by Wongboo** [comment]
@dwmw2  I've rebuild `openconnect` after revert the commit you mentioned, everything seems to work fine! thanks! The output message seems to be fine:
```
^CGET https://vpn.pku.edu.cn/dana-na/auth/logout.cgi
SSL negotiation with vpn.pku.edu.cn
Connected to HTTPS on vpn.pku.edu.cn with ciphersuite (TLS1.2)-(RSA)-(AES-128-CBC)-(SHA256)
Got HTTP response: HTTP/1.1 302 Moved
Set-Cookie: DSID=<elided>; path=/; expires=Thu, 01 Jan 1970 22:00:00 GMT; secure
Set-Cookie: DSLastAccess=x; path=/; expires=Thu, 01 Jan 1970 22:00:00 GMT; secure
Set-Cookie: DSFirstAccess=x; path=/; expires=Thu, 01 Jan 1970 22:00:00 GMT; secure
Set-Cookie: DSOSMLOGIN=x; path=/; expires=Thu, 01 Jan 1970 22:00:00 GMT; secure
Set-Cookie: DSHCSTARTED=x; path=/dana-na/; expires=Thu, 01 Jan 1970 22:00:00 GMT; secure
Set-Cookie: DSCheckBrowser=x; path=/; expires=Thu, 01 Jan 1970 22:00:00 GMT; secure
Set-Cookie: DSHCMOBILESECURITY=x; path=/dana-na/; expires=Thu, 01 Jan 1970 22:00:00 GMT; secure
Set-Cookie: DSJSAMInitialized=x; path=/; expires=Thu, 01 Jan 1970 22:00:00 GMT; secure
Set-Cookie: DSLaunchURL=x; path=/; expires=Thu, 01 Jan 1970 22:00:00 GMT; secure
Date: Sat, 07 May 2022 09:31:49 GMT
location: /dana-na/auth/url_default/welcome.cgi?p=logout&c=4&u=useruidf8c94ad3873e284f069c4493f8b68cba10cfa6e9&signinUrl=hTx2YiFpCwABAAAAmcENGU3k7E2V7WwswIIyahrrkjcHc19cM9d1ll14r%2FZs5zw%2F%2F9bgMMLs3WgNd5fy
ds-connection: close
Content-Type: text/html; charset=utf-8
Set-Cookie: DSLastAccess=1651915909; path=/; Secure
Pragma: no-cache
Cache-Control: no-store
Expires: -1
Content-Length: 0
HTTP body length:  (0)
Logout successful.
delete net default: gateway 192.168.1.2
add net default: gateway 192.168.1.2
delete host 222.29.159.224: gateway 192.168.1.2
User cancelled (SIGINT/SIGTERM); exiting.
```

**Comment by dwmw2** [comment]
Thanks for the quick turnaround in testing; even with my newly-minted AWS Mac instance I would have found it hard to test the default route stuff, as I only seem to be able to access it over Legacy IP so I'd lose my connection to it.

So.... this looks like a bug in vpnc-script on OSX which we should fix; it *ought* to be able to cope with that explicit default route, instead of just the implicit one as before. When the VPN server gives us that route, we ought to honour it; not doing so was the topic *(well, **a** topic)* of https://gitlab.com/openconnect/openconnect/-/issues/245

But I suppose you can consider bug-fix to be a 'new' feature, and given that it causes a regression on OSX it probably makes sense for your 9.01 builds to revert it for now, while we fix up vpnc-script to cope.

@alechemy are you able to confirm whether the same revert fixes your issue too, please? Or if you're using GlobalProtect, see commits https://gitlab.com/openconnect/openconnect/-/commit/99ae55aec1408a2905df72394dab99cb6fb41aed (which was in v8.20) and later revert https://gitlab.com/openconnect/openconnect/-/commit/a2b8134edf8e5f8e942dedf105e2813a0824b919 which was in v9.00.

**Comment by Wongboo** [comment]
revert https://gitlab.com/openconnect/openconnect/-/commit/a2b8134edf8e5f8e942dedf105e2813a0824b919, issue exists
```
route: bad address: 
route: writing to routing socket: File exists
add net default: gateway 192.168.1.2: File exists
delete net 162.105.129.27: gateway 222.29.103.109
delete net 162.105.129.26: gateway 222.29.103.109
route: bad address: 
User cancelled (SIGINT/SIGTERM); exiting.
```
Cannot build successfully after revert  https://gitlab.com/openconnect/openconnect/-/commit/99ae55aec1408a2905df72394dab99cb6fb41aed, might because of some conflict

**Comment by dwmw2** [comment]
@Wongboo Thanks, one of those *is* the revert of the other one, just with the conflicts resolved. But those are for the GlobalProtect VPN protocol, while you are using Network Connect. So I don't think it will change anything for you.

**Comment by alechemy** [comment]
> Thanks for the quick turnaround in testing; even with my newly-minted AWS Mac instance I would have found it hard to test the default route stuff, as I only seem to be able to access it over Legacy IP so I'd lose my connection to it.
> 
> So.... this looks like a bug in vpnc-script on OSX which we should fix; it _ought_ to be able to cope with that explicit default route, instead of just the implicit one as before. When the VPN server gives us that route _and_ other routes, we ought to honour it; not doing so was the topic _(well, **a** topic)_ of [gitlab.com/openconnect/openconnect/-/issues/245](https://gitlab.com/openconnect/openconnect/-/issues/245)
> 
> But I suppose you can consider that bug-fix to be a 'new' feature, and given that it causes a regression on OSX it probably makes sense for your 9.01 builds to revert it for now, while we fix up vpnc-script to cope.
> 
> @alechemy are you able to confirm whether the same revert fixes your issue too, please? Or if you're using GlobalProtect, see commits [gitlab.com/openconnect/openconnect/-/commit/99ae55aec1408a2905df72394dab99cb6fb41aed](https://gitlab.com/openconnect/openconnect/-/commit/99ae55aec1408a2905df72394dab99cb6fb41aed) (which was in v8.20) and later revert [gitlab.com/openconnect/openconnect/-/commit/a2b8134edf8e5f8e942dedf105e2813a0824b919](https://gitlab.com/openconnect/openconnect/-/commit/a2b8134edf8e5f8e942dedf105e2813a0824b919) which was in v9.00.

Hi @dwmw2 . Thanks for looping me in. I've managed to revert the commit you linked and have reinstalled `openconnect` with that change. 

I can gladly confirm that this change looks to have resolved the issue for me. I don't have time at this moment to collect detailed logs, but here's some quick info on my situation, in case it helps:

- I typically use `openconnect` in this way:
  `sudo openconnect --protocol=nc --no-dtls --cookie "DSID=X" "vpn.x.com"`
- Yesterday my machine auto-upgraded to 9.0.1, and I started having two issues: 
  1) Upon connection, I could access "public" sites just fine, but ones that were gated behind the VPN did not load. 
  2) Upon closing (via ctrl+c), I got the `route: bad address` log mentioned above, and my WiFi connection was left in a bad state (I couldn't load any pages, public or otherwise), until I toggled my connection off and on again.

But yeah, as mentioned, reverting the commit you specified seems to have cleared up both issues for me. 🎉

**Comment by dwmw2** [comment]
Thanks for the confirmation. So we should really fix vpnc-script to cope with the explicit 0.0.0.0/0 route, which *can* happen in some configurations anyway; the commit you're reverting just made it more common.

**Comment by SMillerDev** [comment]
Closing since it's an upstream issue
